// @auth0/cosmos/atoms/radio/radio.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var automation_attribute_1 = __importDefault(require("../../_helpers/automation-attribute"));
var form_1 = __importDefault(require("../../molecules/form"));
var styled_1 = __importDefault(require("../../styled"));
var tokens_1 = require("../../tokens");
var CheckMark = styled_1.default.span(templateObject_1 || (templateObject_1 = __makeTemplateObject([""], [""])));
var Label = styled_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
var justifyContent = {
    horizontal: "margin-right: " + tokens_1.spacing.medium,
    vertical: "margin-bottom: " + tokens_1.spacing.xsmall
};
var Radio = function (props) { return (React.createElement(Radio.Element, __assign({}, props, automation_attribute_1.default("radio")), React.Children.map(props.children, function (child) {
    if (!child) {
        return null;
    }
    return React.cloneElement(child, {
        name: props.name,
        checked: props.selected === child.props.value,
        readOnly: props.readOnly || child.props.readOnly
    });
}))); };
var RadioOption = function (_a) {
    var readOnly = _a.readOnly, children = _a.children, props = __rest(_a, ["readOnly", "children"]);
    return (React.createElement(RadioOption.Element, { readOnly: readOnly },
        React.createElement(form_1.default.Field.ContextConsumer, null, function (context) { return (React.createElement("input", __assign({ id: props.id || context.formFieldId }, automation_attribute_1.default("radio.option"), { type: "radio", readOnly: true, disabled: readOnly }, props))); }),
        React.createElement(CheckMark, null),
        React.createElement(Label, null, children)));
};
RadioOption.Element = styled_1.default.label(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  position: relative;\n  cursor: ", ";\n  padding-left: ", ";\n  user-select: none;\n  vertical-align: middle;\n\n  input {\n    position: absolute;\n    opacity: 0;\n  }\n\n  ", " {\n    position: absolute;\n    top: 0;\n    left: 0;\n    transform: translateY(20%);\n    height: 16px;\n    width: 16px;\n    background-color: ", ";\n    border: 1px solid ", ";\n    box-shadow: ", ";\n    border-radius: 50%;\n  }\n\n  &:hover input ~ ", " {\n    box-shadow: ", ";\n  }\n\n  input:checked ~ ", " {\n    background-color: ", ";\n    border: 1px solid ", ";\n  }\n\n  input:focus ~ ", " {\n    box-shadow: inset 0 1px 4px 0 ", ", 0 0 0 2px ", ";\n  }\n\n  ", ":after {\n    content: \"\";\n    position: absolute;\n    display: none;\n  }\n\n  input:checked ~ ", ":after {\n    display: block;\n  }\n\n  ", ":after {\n    background-color: ", ";\n    width: 6px;\n    height: 6px;\n    border-radius: 50%;\n    box-shadow: 0 1px 2px 0 ", ";\n    left: 50%;\n    top: 50%;\n    transform: translate(-50%, -50%);\n  }\n"], ["\n  position: relative;\n  cursor: ", ";\n  padding-left: ", ";\n  user-select: none;\n  vertical-align: middle;\n\n  input {\n    position: absolute;\n    opacity: 0;\n  }\n\n  ", " {\n    position: absolute;\n    top: 0;\n    left: 0;\n    transform: translateY(20%);\n    height: 16px;\n    width: 16px;\n    background-color: ", ";\n    border: 1px solid ", ";\n    box-shadow: ", ";\n    border-radius: 50%;\n  }\n\n  &:hover input ~ ", " {\n    box-shadow: ", ";\n  }\n\n  input:checked ~ ", " {\n    background-color: ", ";\n    border: 1px solid ", ";\n  }\n\n  input:focus ~ ", " {\n    box-shadow: inset 0 1px 4px 0 ", ", 0 0 0 2px ", ";\n  }\n\n  ", ":after {\n    content: \"\";\n    position: absolute;\n    display: none;\n  }\n\n  input:checked ~ ", ":after {\n    display: block;\n  }\n\n  ", ":after {\n    background-color: ", ";\n    width: 6px;\n    height: 6px;\n    border-radius: 50%;\n    box-shadow: 0 1px 2px 0 ", ";\n    left: 50%;\n    top: 50%;\n    transform: translate(-50%, -50%);\n  }\n"])), function (props) { return (props.readOnly ? "not-allowed" : "pointer"); }, tokens_1.spacing.medium, CheckMark, function (props) { return (props.readOnly ? tokens_1.colors.radio.backgroundDisabled : tokens_1.colors.radio.background); }, function (props) { return (props.readOnly ? tokens_1.colors.radio.borderDisabled : tokens_1.colors.radio.border); }, function (props) { return (props.readOnly ? "none" : "inset 0 1px 2px 0  " + tokens_1.colors.radio.shadow); }, CheckMark, function (props) { return (props.readOnly ? "none" : "inset 0 1px 4px 0 " + tokens_1.colors.radio.shadow); }, CheckMark, function (props) { return (props.readOnly ? tokens_1.colors.base.gray : tokens_1.colors.radio.backgroundSelected); }, function (props) { return (props.readOnly ? tokens_1.colors.base.gray : tokens_1.colors.radio.borderSelected); }, CheckMark, tokens_1.colors.radio.shadow, tokens_1.colors.base.blue, CheckMark, CheckMark, CheckMark, tokens_1.colors.base.white, tokens_1.colors.radio.shadow);
Radio.Option = RadioOption;
Radio.Element = styled_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", " {\n    display: ", ";\n    ", ";\n\n    &:last-child {\n      margin: 0;\n    }\n  }\n"], ["\n  ", " {\n    display: ", ";\n    ", ";\n\n    &:last-child {\n      margin: 0;\n    }\n  }\n"])), Radio.Option.Element, function (props) { return (props.align === "horizontal" ? "inline-block" : "table"); }, function (props) { return justifyContent[props.align]; });
Radio.defaultProps = {
    align: "vertical"
};
var StyledRadio = Radio.Element;
exports.StyledRadio = StyledRadio;
var StyledRadioOption = Radio.Option.Element;
exports.StyledRadioOption = StyledRadioOption;
exports.default = Radio;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
