// @auth0/cosmos/atoms/select/select.story.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var react_1 = require("@storybook/react");
var __1 = require("../../");
var story_helpers_1 = require("../../_helpers/story-helpers");
var styled_1 = __importDefault(require("../../styled"));
var ExampleForSelect = styled_1.default(story_helpers_1.Example)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  min-height: 450px;\n"], ["\n  min-height: 450px;\n"])));
var InteractiveSelect = (function (_super) {
    __extends(InteractiveSelect, _super);
    function InteractiveSelect(props) {
        var _this = _super.call(this, props) || this;
        _this.state = { value: null };
        _this.onChange = _this.onChange.bind(_this);
        return _this;
    }
    InteractiveSelect.prototype.onChange = function (event) {
        this.setState({ value: event.target.value });
    };
    InteractiveSelect.prototype.render = function () {
        return this.props.children(this.state.value, this.onChange);
    };
    return InteractiveSelect;
}(React.Component));
react_1.storiesOf("Select", module).add("simple", function () { return (React.createElement(ExampleForSelect, { title: "Select: simple" },
    React.createElement(__1.Select, { options: [{ text: "One", value: 1 }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("simple with placeholder", function () { return (React.createElement(ExampleForSelect, { title: "Select: with placeholder" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement(__1.Select, { placeholder: "Select an option...", value: value, options: [{ text: "One", value: 1 }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: onChange })); }))); });
react_1.storiesOf("Select", module).add("simple with error", function () { return (React.createElement(ExampleForSelect, { title: "Select: simple" },
    React.createElement(__1.Select, { hasError: true, options: [{ text: "One", value: 1 }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("multiple tags", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed multiple tags" },
    React.createElement(__1.Select, { multiple: true, value: [1, 2, 3], options: [{ text: "Applications", value: 1 }, { text: "Connections", value: 2 }, { text: "Logs", value: 3 }] }))); });
react_1.storiesOf("Select", module).add("simple with custom getOptionValue", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom getOptionValue" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement("div", null,
        React.createElement(__1.Select, { options: [
                { text: "1", myCustomValue: 1 },
                { text: "2", myCustomValue: 2 },
                { text: "3", myCustomValue: 3 },
                { text: "4", myCustomValue: 4 },
                { text: "5", myCustomValue: 5 },
                { text: "6", myCustomValue: 6 },
                { text: "7", myCustomValue: 7 },
                { text: "8", myCustomValue: 8 },
                { text: "9", myCustomValue: 9 },
                { text: "10", myCustomValue: 10 },
                { text: "11", myCustomValue: 11 },
                { text: "12", myCustomValue: 12 },
                { text: "13", myCustomValue: 13 },
                { text: "14", myCustomValue: 14 },
                { text: "15", myCustomValue: 15 },
                { text: "16", myCustomValue: 16 },
                { text: "17", myCustomValue: 17 },
                { text: "18", myCustomValue: 18 }
            ], value: value, onChange: onChange, getOptionValue: function (option) { return option.myCustomValue; } }),
        React.createElement("p", null,
            "Selected value: ",
            value))); }))); });
react_1.storiesOf("Select", module).add("multiple with custom getOptionValue", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom getOptionValue" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement("div", null,
        React.createElement(__1.Select, { multiple: true, searchable: true, options: [
                { text: "One", myCustomValue: 1 },
                { text: "Two", myCustomValue: 2 },
                { text: "Three", myCustomValue: 3 }
            ], value: value, onChange: onChange, getOptionValue: function (option) { return option.myCustomValue; } }),
        React.createElement("p", null,
            "Selected value: ",
            (value || []).join(",")))); }))); });
react_1.storiesOf("Select", module).add("multiple tags: open", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed multiple tags" },
    React.createElement(__1.Select, { multiple: true, defaultMenuOpen: true, value: [1], options: [{ text: "Applications", value: 1 }, { text: "Connections", value: 2 }, { text: "Logs", value: 3 }] }))); });
react_1.storiesOf("Select", module).add("stressed: multiple tags", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed multiple tags" },
    React.createElement(__1.Select, { multiple: true, defaultMenuOpen: true, value: [1, 2, 3, 4, 5, 6, 7], options: [
            { text: "This is a very long tag, like super long", value: 1 },
            { text: "This is a very long tag, like super long for real", value: 2 },
            { text: "This is a long tag, like super long for real", value: 3 },
            { text: "Very long tag, like super long for real", value: 4 },
            { text: "Long tag, like super long for real", value: 5 },
            { text: "This is a long tag", value: 6 },
            { text: "This is a very long tag, like super long for real", value: 7 }
        ] }))); });
react_1.storiesOf("Select", module).add("stressed: multiple tags - open", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed multiple tags" },
    React.createElement(__1.Select, { multiple: true, defaultMenuOpen: true, value: [1], options: [
            { text: "This is a very long tag, like super long", value: 1 },
            { text: "This is a very long tag, like super long for real", value: 2 },
            { text: "This is a long tag, like super long for real", value: 3 },
            { text: "Very long tag, like super long for real", value: 4 },
            { text: "Long tag, like super long for real", value: 5 },
            { text: "This is a long tag", value: 6 },
            { text: "This is a very long tag, like super long for real", value: 7 }
        ] }))); });
var SelectWithEmptyString = (function (_super) {
    __extends(SelectWithEmptyString, _super);
    function SelectWithEmptyString(props) {
        var _this = _super.call(this, props) || this;
        _this.state = { value: "custom" };
        _this.handleValue = _this.handleValue.bind(_this);
        return _this;
    }
    SelectWithEmptyString.prototype.handleValue = function (event) {
        this.setState({ value: event.target.value });
    };
    SelectWithEmptyString.prototype.render = function () {
        return (React.createElement(__1.Select, { defaultMenuOpen: this.props.defaultMenuOpen, options: [
                { text: "All apps", value: "" },
                { text: "Custom apps", value: "custom" },
                { text: "Private apps", value: "private" }
            ], value: this.state.value, onChange: this.handleValue }));
    };
    return SelectWithEmptyString;
}(React.Component));
react_1.storiesOf("Select", module).add("simple with empty string", function () { return (React.createElement(ExampleForSelect, { title: "Select: simple" },
    React.createElement(SelectWithEmptyString, null))); });
react_1.storiesOf("Select", module).add("default value", function () { return (React.createElement(ExampleForSelect, { title: "Select: with default value" },
    React.createElement(__1.Select, { value: 2, options: [{ text: "One", value: 1 }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("disabled", function () { return (React.createElement(ExampleForSelect, { title: "Select: disabled" },
    React.createElement(__1.Select, { value: 2, disabled: true, options: [{ text: "One", value: 1 }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("disabled option", function () { return (React.createElement(ExampleForSelect, { title: "Select: disabled option" },
    React.createElement(__1.Select, { defaultMenuOpen: true, placeholder: "Select from the enabled options", options: [{ text: "One", value: 1, disabled: true }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("disabled with placeholder", function () { return (React.createElement(ExampleForSelect, { title: "Select: disabled with placeholder" },
    React.createElement(__1.Select, { defaultMenuOpen: true, disabled: true, placeholder: "Select from the enabled options", options: [{ text: "One", value: 1, disabled: true }, { text: "Two", value: 2 }, { text: "Three", value: 3 }], onChange: function (event) { return console.log(event); } }))); });
react_1.storiesOf("Select", module).add("groups", function () { return (React.createElement(ExampleForSelect, { title: "Select: groups" },
    React.createElement(__1.Select, { defaultMenuOpen: true, placeholder: "Groups", options: [
            {
                groupName: "Group one",
                items: [{ text: "One one", value: 11 }, { text: "One two", value: 12 }, { text: "One three", value: 13 }]
            },
            {
                groupName: "Group two",
                items: [{ text: "Two one", value: 21 }, { text: "Two two", value: 22 }, { text: "Two three", value: 23 }]
            }
        ] }))); });
react_1.storiesOf("Select", module).add("interactive: groups", function () { return (React.createElement(ExampleForSelect, { title: "Select: groups" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement(__1.Select, { value: value, onChange: onChange, placeholder: "Groups", options: [
            {
                groupName: "Group one",
                items: [{ text: "One one", value: 11 }, { text: "One two", value: 12 }, { text: "One three", value: 13 }]
            },
            {
                groupName: "Group two",
                items: [{ text: "Two one", value: 21 }, { text: "Two two", value: 22 }, { text: "Two three", value: 23 }]
            }
        ] })); }))); });
react_1.storiesOf("Select", module).add("interactive: groups - multiple", function () { return (React.createElement(ExampleForSelect, { title: "Select: groups" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement(__1.Select, { multiple: true, searchable: true, value: value, onChange: onChange, placeholder: "Groups", options: [
            {
                groupName: "Group one",
                items: [{ text: "One one", value: 11 }, { text: "One two", value: 12 }, { text: "One three", value: 13 }]
            },
            {
                groupName: "Group two",
                items: [{ text: "Two one", value: 21 }, { text: "Two two", value: 22 }, { text: "Two three", value: 23 }]
            }
        ] })); }))); });
react_1.storiesOf("Select", module).add("stressed", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed - 56 characters with 300px width" },
    React.createElement("div", { style: { width: 300 } },
        React.createElement(__1.Select, { value: 1, defaultMenuOpen: true, options: [
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 1
                },
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 2
                },
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 3
                }
            ], onChange: function (event) { return console.log(event); } })))); });
react_1.storiesOf("Select", module).add("interactive: stressed", function () { return (React.createElement(ExampleForSelect, { title: "Select: stressed - 56 characters with 300px width" },
    React.createElement("div", { style: { width: 300 } },
        React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement(__1.Select, { value: value, onChange: onChange, defaultMenuOpen: true, options: [
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 1
                },
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 2
                },
                {
                    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    value: 3
                }
            ] })); })))); });
var CustomRendererExample = (function (_super) {
    __extends(CustomRendererExample, _super);
    function CustomRendererExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = { selectedItems: [] };
        return _this;
    }
    CustomRendererExample.prototype.handleChange = function (event) {
        var selectedItems = event.target.value;
        this.setState({ selectedItems: selectedItems });
    };
    CustomRendererExample.prototype.renderOption = function (option, _a) {
        var isHovered = _a.isHovered;
        return (React.createElement("div", null,
            React.createElement("p", null,
                React.createElement("strong", null, option.label)),
            React.createElement("p", null, option.description)));
    };
    CustomRendererExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(__1.Select, { value: this.state.selectedItems, onChange: function (ev) { return _this.handleChange(ev); }, placeholder: "Select an item...", customOptionRenderer: this.renderOption, defaultMenuOpen: this.props.defaultMenuOpen, options: [
                {
                    label: "One",
                    description: "This item holds the One number as 1.",
                    value: 1
                },
                {
                    label: "Two",
                    description: "This item holds the Two number as 2.",
                    value: 2
                },
                {
                    label: "Three",
                    description: "This item holds the Three number as 3.",
                    value: 3
                },
                {
                    label: "Four",
                    description: "This item holds the Four number as 4.",
                    value: 4
                },
                {
                    label: "Five",
                    description: "This item holds the Five number as 5.",
                    value: 5
                },
                {
                    label: "Six",
                    description: "This item holds the Six number as 6.",
                    value: 6
                },
                {
                    label: "Seven",
                    description: "This item holds the Seven number as 7.",
                    value: 7
                },
                {
                    label: "Eight",
                    description: "This item holds the Eight number as 8.",
                    value: 8
                },
                {
                    label: "Nine",
                    description: "This item holds the Nine number as 9.",
                    value: 9
                },
                {
                    label: "Ten",
                    description: "This item holds the Ten number as 10.",
                    value: 10
                }
            ] }));
    };
    return CustomRendererExample;
}(React.Component));
react_1.storiesOf("Select", module).add("custom renderer", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom renderer" },
    React.createElement(CustomRendererExample, null))); });
react_1.storiesOf("Select", module).add("custom value renderer", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom value renderer" },
    React.createElement(__1.Select, { defaultMenuOpen: true, value: "Harry Kane", onChange: function (ev) { return _this.handleChange(ev); }, placeholder: "Select a custom item...", customValueRenderer: function (option) { return (React.createElement("span", null,
            option.country,
            " ",
            option.label)); }, options: [
            {
                value: "Harry Kane",
                text: "Harry Kane",
                country: "🇬🇧"
            },
            {
                value: "Lionel Messi",
                text: "Lionel Messi",
                country: "🇦🇷"
            },
            {
                value: "Antoine Griezmann",
                text: "Antoine Griezmann",
                country: "🇫🇷"
            },
            {
                value: "Luis Suarez",
                text: "Luis Suarez",
                country: "🇺🇾"
            }
        ] }))); });
react_1.storiesOf("Select", module).add("interactive: custom value renderer", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom value renderer (interactive)" },
    React.createElement(InteractiveSelect, null, function (value, onChange) { return (React.createElement(__1.Select, { value: value, onChange: onChange, placeholder: "Select a custom item...", customValueRenderer: function (option) { return (React.createElement("span", null,
            option.country,
            " ",
            option.label)); }, options: [
            {
                value: "Harry Kane",
                text: "Harry Kane",
                country: "🇬🇧"
            },
            {
                value: "Lionel Messi",
                text: "Lionel Messi",
                country: "🇦🇷"
            },
            {
                value: "Antoine Griezmann",
                text: "Antoine Griezmann",
                country: "🇫🇷"
            },
            {
                value: "Luis Suarez",
                text: "Luis Suarez",
                country: "🇺🇾"
            }
        ] })); }))); });
react_1.storiesOf("Select", module).add("custom renderer in dialog", function () { return (React.createElement(ExampleForSelect, { title: "Select: custom renderer" },
    React.createElement(__1.Dialog, { open: true, title: "A dialog example" },
        React.createElement(CustomRendererExample, { defaultMenuOpen: true })))); });
var AsyncExample = (function (_super) {
    __extends(AsyncExample, _super);
    function AsyncExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = { selectedItems: [] };
        _this.allOptions = [
            {
                label: "One",
                description: "This item holds the One number as 1.",
                value: 1
            },
            {
                label: "Two",
                description: "This item holds the Two number as 2.",
                value: 2
            },
            {
                label: "Three",
                description: "This item holds the Three number as 3.",
                value: 3
            },
            {
                label: "Four",
                description: "This item holds the Four number as 4.",
                value: 4
            },
            {
                label: "Five",
                description: "This item holds the Five number as 5.",
                value: 5
            },
            {
                label: "Six",
                description: "This item holds the Six number as 6.",
                value: 6
            },
            {
                label: "Seven",
                description: "This item holds the Seven number as 7.",
                value: 7
            },
            {
                label: "Eight",
                description: "This item holds the Eight number as 8.",
                value: 8
            },
            {
                label: "Nine",
                description: "This item holds the Nine number as 9.",
                value: 9
            },
            {
                label: "Ten",
                description: "This item holds the Ten number as 10.",
                value: 10
            }
        ];
        _this.loadOptions = _this.loadOptions.bind(_this);
        _this.noOptionsMessage = _this.noOptionsMessage.bind(_this);
        return _this;
    }
    AsyncExample.prototype.handleChange = function (event) {
        var selectedItems = event.target.value;
        console.log({ selectedItems: selectedItems });
        this.setState({ selectedItems: selectedItems });
    };
    AsyncExample.prototype.filterOptions = function (value) {
        return this.allOptions.filter(function (item) { return item.label.toLowerCase().includes(value.toLowerCase()); });
    };
    AsyncExample.prototype.loadOptions = function (inputValue, callback) {
        var _this = this;
        setTimeout(function () {
            callback(_this.filterOptions(inputValue));
        }, 500);
    };
    AsyncExample.prototype.noOptionsMessage = function (_a) {
        var inputValue = _a.inputValue;
        return inputValue.length > 0 ? "No options" : "Type to search...";
    };
    AsyncExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(__1.Select, { async: true, value: this.state.selectedItems, onChange: function (ev) { return _this.handleChange(ev); }, placeholder: "Select an item...", noOptionsMessage: this.noOptionsMessage, loadOptions: this.loadOptions }));
    };
    return AsyncExample;
}(React.Component));
react_1.storiesOf("Select", module).add("async select", function () { return (React.createElement(ExampleForSelect, { title: "Select: async select" },
    React.createElement(AsyncExample, null))); });
react_1.storiesOf("Select", module).add("disabled options", function () { return (React.createElement(ExampleForSelect, { title: "Select: disabled options" },
    React.createElement("div", { style: { width: 300 } },
        React.createElement(__1.Select, { value: 3, searchable: true, defaultMenuOpen: true, options: [
                {
                    text: "Lorem ipsum dolor sit amet",
                    value: 1,
                    disabled: true
                },
                {
                    text: "Lorem ipsum dolor sit amet",
                    value: 2,
                    disabled: true
                },
                {
                    text: "Lorem ipsum dolor sit amet",
                    value: 3
                }
            ], onChange: function (event) { return console.log(event); } })))); });
var templateObject_1;
