// @auth0/cosmos/_helpers/action-shape.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var PropTypes = __importStar(require("prop-types"));
var icon_1 = require("../atoms/icon");
var shapeForDocs = { label: 'string', icon: 'enum', handler: 'func' };
exports.shapeForDocs = shapeForDocs;
var actionShape = PropTypes.shape({
    label: PropTypes.string.isRequired,
    icon: PropTypes.oneOf(icon_1.__ICONNAMES__),
    handler: PropTypes.func.isRequired
});
exports.actionShape = actionShape;
var actionShapeWithoutRequiredHandler = PropTypes.shape({
    label: PropTypes.string.isRequired,
    icon: PropTypes.oneOf(icon_1.__ICONNAMES__),
    handler: PropTypes.func
});
exports.actionShapeWithoutRequiredHandler = actionShapeWithoutRequiredHandler;
var actionShapeWithRequiredIcon = PropTypes.shape({
    label: PropTypes.string.isRequired,
    icon: PropTypes.oneOf(icon_1.__ICONNAMES__).isRequired,
    handler: PropTypes.func.isRequired
});
exports.actionShapeWithRequiredIcon = actionShapeWithRequiredIcon;
